### PR TITLE
Add performance.{now,timeOrigin}

### PR DESCRIPTION
**Pull request by saghul**
Ref: https://github.com/quickjs-ng/quickjs/issues/16

### CODE REVIEW

**Comment by saghul** [comment]
@bnoordhuis Is this what you had in mind or would you rather not add it to the default context? Maybe a compile time flag? A new `JS_NewContext2(rt, flags)` ?

File: quickjs.c [hunk 1]
```diff
@@ -50219,3 +50224,37 @@ void JS_AddIntrinsicTypedArrays(JSContext *ctx)
...
 }
+
+/* Performance */
+
+static uint64_t js__now_ms(void)
+{
+    // TODO(saghul) Windows support.
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
+    return (((uint64_t) ts.tv_sec) * 1000 + ts.tv_nsec / 1e6);
```
Comment (bnoordhuis) [inline quickjs.c:50235]:
```suggestion
    return (uint64_t) ts.tv_sec * 1000 + ts.tv_nsec / 1e6;
```
You could also write it as `return ts.tv_sec * 1000ull + ts.tv_nsec / 1e6;`

File: quickjs.c [hunk 2]
```diff
@@ -50219,3 +50224,37 @@ void JS_AddIntrinsicTypedArrays(JSContext *ctx)
...
+    ctx->time_origin = js__now_ms();
+
+    JSValue performance = JS_NewObject(ctx);
+    JS_SetPropertyFunctionList(ctx, performance, js_perf_proto_funcs, countof(js_perf_proto_funcs));
+    JS_DefinePropertyValueStr(ctx, performance, "timeOrigin",
+                           JS_NewFloat64(ctx, ctx->time_origin),
+                           JS_PROP_ENUMERABLE);
+    JS_DefinePropertyValueStr(ctx, ctx->global_obj, "performance",
+                           JS_DupValue(ctx, performance),
+                           JS_PROP_ENUMERABLE);
```
Comment (bnoordhuis) [inline quickjs.c:50258]:
```suggestion
    JS_DefinePropertyValueStr(ctx, performance, "timeOrigin",
                              JS_NewFloat64(ctx, ctx->time_origin),
                              JS_PROP_ENUMERABLE);
    JS_DefinePropertyValueStr(ctx, ctx->global_obj, "performance",
                              JS_DupValue(ctx, performance),
                              JS_PROP_ENUMERABLE);
```

**Review by bnoordhuis — APPROVED** [review]
Nice!

**Comment by bnoordhuis** [comment]
> Is this what you had in mind or would you rather not add it to the default context? Maybe a compile time flag? A new JS_NewContext2(rt, flags) ?

I was thinking of an opt-in (so not adding it by default) but, OTOH, it's probably fine.

**Pull request merged**